include/ygm/io/multi_output.hpp

```diff
@@ -0,0 +1,121 @@
+
+// Copyright 2019-2021 Lawrence Livermore National Security, LLC and other YGM
+// Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <ygm/container/detail/hash_partitioner.hpp>
+
+namespace ygm::io {
+
+namespace fs = std::filesystem;
+
+template <typename Partitioner =
+              ygm::container::detail::hash_partitioner<std::string>>
+class multi_output {
+ public:
+  using self_type = multi_output<Partitioner>;
+
+  multi_output(ygm::comm &comm, const std::string &filename_prefix,
+               bool append = false)
+      : m_comm(comm),
+        pthis(this),
+        m_prefix_path(filename_prefix),
+        m_append_flag(append) {
+    pthis.check(m_comm);
+
+    fs::path prefix_path(filename_prefix);
+
+    // Create directories to hold files
+    if (comm.rank0()) {
+      make_directories(m_prefix_path);
+    }
+  }
+
+  ~multi_output() {
+    m_comm.barrier();
+    for (const auto &[fname, file_ptr] : m_map_file_pointers) {
+      file_ptr->close();
+    }
+  }
+
+  template <typename... Args>
+  void async_write_line(const std::string &filename, Args &&... args) {
+    std::string s = pack_stream(args...);
+
+    m_comm.async(
+        owner(filename),
+        [](auto mo_ptr, const std::string &filename, const std::string &s) {
+          fs::path fullname(mo_ptr->m_prefix_path);
+          fullname += filename;
+
+          auto ofstream_iter = mo_ptr->m_map_file_pointers.find(filename);
+          if (ofstream_iter == mo_ptr->m_map_file_pointers.end()) {
+            const auto [iter, success] = mo_ptr->m_map_file_pointers.insert(
+                std::make_pair(filename, mo_ptr->make_ofstream_ptr(fullname)));
+            ofstream_iter = iter;
+          }
+
+          *(ofstream_iter->second) << s << "\n";
+        },
+        pthis, filename, s);
+  }
+
+  ygm::comm &comm() { return m_comm; }
+
+ private:
+  int owner(const std::string &filename) {
+    auto [owner, bank] = partitioner(filename, m_comm.size(), 1024);
+    return owner;
+  }
+
+  template <typename... Args>
+  std::string pack_stream(Args &&... args) const {
+    std::stringstream ss;
+    (ss << ... << args);
+    return ss.str();
+  }
+
+  void make_directories(const fs::path &p) {
```

Comment:
if they pass a file (not directory) that already exists, need to error (I think).